File: python-elasticemail-examples/examples/with_template.py
"""
Template Example

Templates are referenced by name. The template is created on first run,
then a transactional email is sent with merge values.

Usage: python examples/with_template.py
"""

import os
import sys

import ElasticEmail

sys.path.insert(0, os.path.dirname(__file__))
from ee import FROM, TEMPLATE_NAME, TO, get_configuration, print_api_error


def ensure_template(templates_api):
    try:
        templates_api.templates_by_name_get(TEMPLATE_NAME)
        print('Template "{}" already exists.'.format(TEMPLATE_NAME))
        return
    except ElasticEmail.ApiException as e:
        if e.status != 404:
            raise

    templates_api.templates_post(
        ElasticEmail.TemplatePayload(
            Name=TEMPLATE_NAME,
            Subject="Welcome, {firstname}!",
            Body=[
                ElasticEmail.BodyPart(
                    ContentType="HTML",
                    Content="<h1>Welcome, {firstname}!</h1><p>Thanks for joining {company}.</p>",
                ),
            ],
            TemplateScope="Personal",
        )
    )
    print('Template "{}" created.'.format(TEMPLATE_NAME))


def main():
    configuration = get_configuration()

    with ElasticEmail.ApiClient(configuration) as api_client:
        emails_api = ElasticEmail.EmailsApi(api_client)
        templates_api = ElasticEmail.TemplatesApi(api_client)

        try:
            ensure_template(templates_api)

            # Merge values replace {placeholders} in the template subject and body.
            message = ElasticEmail.EmailTransactionalMessageData(
                Recipients=ElasticEmail.TransactionalRecipient(To=[TO]),
                Content=ElasticEmail.EmailContent(
                    From=FROM,
                    TemplateName=TEMPLATE_NAME,
                    Merge={"firstname": "Ann", "company": "Acme"},
                ),
            )
            result = emails_api.emails_transactional_post(message)
            print("Template email sent successfully!")
            print("Transaction ID:", result.transaction_id)
            print("Message ID:", result.message_id)
        except ElasticEmail.ApiException as e:
            print_api_error("template send", e)
            sys.exit(1)


if __name__ == "__main__":
    main()

File: ruby-elasticemail-examples/examples/with_template.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

##
# Using Templates Example
#
# Templates are referenced by name. The template is created on first run,
# then a transactional email is sent with merge values.
#
# Usage: ruby examples/with_template.rb

require_relative "ee"

emails_api = ElasticEmail::EmailsApi.new
templates_api = ElasticEmail::TemplatesApi.new

def ensure_template(templates_api)
  templates_api.templates_by_name_get(TEMPLATE_NAME)
  puts "Template \"#{TEMPLATE_NAME}\" already exists."
rescue ElasticEmail::ApiError => e
  raise e unless e.code == 404

  templates_api.templates_post(
    ElasticEmail::TemplatePayload.new(
      name: TEMPLATE_NAME,
      subject: "Welcome, {firstname}!",
      body: [
        ElasticEmail::BodyPart.new(
          content_type: "HTML",
          content: "<h1>Welcome, {firstname}!</h1><p>Thanks for joining {company}.</p>"
        )
      ],
      template_scope: "Personal"
    )
  )
  puts "Template \"#{TEMPLATE_NAME}\" created."
end

begin
  ensure_template(templates_api)

  # Merge values replace {placeholders} in the template subject and body.
  result = emails_api.emails_transactional_post(
    ElasticEmail::EmailTransactionalMessageData.new(
      recipients: ElasticEmail::TransactionalRecipient.new(to: [TO]),
      content: ElasticEmail::EmailContent.new(
        from: FROM,
        template_name: TEMPLATE_NAME,
        merge: { "firstname" => "Ann", "company" => "Acme" }
      )
    )
  )

  puts "Template email sent successfully!"
  puts "Transaction ID: #{result.transaction_id}"
  puts "Message ID: #{result.message_id}"
rescue ElasticEmail::ApiError => e
  api_error("template send", e)
end
